cashapppay/src/main/java/com/adyen/checkout/cashapppay/DefaultCashAppPayDelegate.kt

```diff
@@ -0,0 +1,231 @@
+import com.adyen.checkout.components.util.PaymentMethodTypes
+import com.adyen.checkout.core.exception.ComponentException
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+@Suppress("TooManyFunctions")
+internal class DefaultCashAppPayDelegate(
+    private val paymentMethod: PaymentMethod,
+    private val configuration: CashAppPayConfiguration
+) : CashAppPayDelegate {
+
+    override fun getPaymentMethodType(): String {
+        return paymentMethod.type ?: PaymentMethodTypes.UNKNOWN
+    }
+
+    private lateinit var cashAppPay: CashAppPay
+
+    override fun shouldCreateComponentStateOnInit(): Boolean = false
+
+    override suspend fun initialize(outputData: CashAppPayOutputData?, onCashAppPayStateChanged: (CashAppPayState) -> Unit) {
+        val cashAppParams = getCashAppParams()
+        cashAppPay = initCashAppPay(cashAppParams, onCashAppPayStateChanged)
+        if (!requiresInput()) {
+            initiateCashAppPayment(cashAppParams, outputData)
+        }
+    }
+
+    @Suppress("ThrowsCount")
+    private fun getCashAppParams(): CashAppParams {
+        return CashAppParams(
+            clientId = paymentMethod.configuration?.clientId
+                ?: throw ComponentException("Cannot launch Cash App Pay, clientId is missing from the payment method object"),
+            scopeId = paymentMethod.configuration?.scopeId
+                ?: throw ComponentException("Cannot launch Cash App Pay, scopeId is missing from the payment method object"),
+            returnUrl = configuration.returnUrl
+                ?: throw ComponentException("Cannot launch Cash App Pay, set the returnUrl in your CashAppPayConfiguration.Builder"),
+        )
+    }
+
+    private fun initCashAppPay(cashAppParams: CashAppParams, onCashAppPayStateChanged: (CashAppPayState) -> Unit): CashAppPay {
+        return when (configuration.cashAppPayEnvironment) {
+            CashAppPayEnvironment.SANDBOX -> CashAppPayFactory.createSandbox(cashAppParams.clientId)
+            CashAppPayEnvironment.PRODUCTION -> CashAppPayFactory.create(cashAppParams.clientId)
+        }.apply {
+            registerForStateUpdates(DefaultCashAppPayListener(onCashAppPayStateChanged))
+        }
+    }
+
+    private suspend fun initiateCashAppPayment(cashAppParams: CashAppParams, outputData: CashAppPayOutputData?) {
+        val actions = listOfNotNull(
+            getOneTimeAction(cashAppParams),
+            getOnFileAction(cashAppParams, outputData),
+        )
+
+        if (actions.isEmpty()) {
+            throw ComponentException("Cannot launch Cash App Pay, you need to either pass an amount or store the shopper account")
+        }
+
+        withContext(Dispatchers.IO) {
+            // must be called from a background thread according to Cash App Pay docs
+            cashAppPay.createCustomerRequest(actions, cashAppParams.returnUrl)
```

Comment:
Does this actually do IO work? If not, the default dispatcher would be better

Reply:
This is from the Cash App SDK, I think we should explicitly specify `Dispatchers.IO` right?
```kotlin
  /**
   * Create customer request given list of [CashAppPayPaymentAction].
   *
   * Must be called from a background thread.
   * ...
   */
  @WorkerThread
  fun createCustomerRequest(paymentActions: List<CashAppPayPaymentAction>, redirectUri: String?)
```

Reply:
Checked the code and it does a network call, so `IO` is perfect 👍